Verify VCs and VPs

simne7 opened the pull request.

simne7 committed:
feat(verification-service): verify linked presentations for each did doc

simne7 committed:
fix(verification-service): fix dependency imports

jceb commented on services/verification-service/src/verifications/verification_service.rs:
```diff
@@ -0,0 +1,295 @@
+    }
+}
+
+fn map_ssi_error_to_verification_result(error: ssi::claims::Invalid) -> VcVerificationResult {
```
I asked Claude about associating the enum with a static string/error message (and the next example including an error code). It came up with the following for statically associating an error message with an enum:

```rust
// Method 1: Static messages in Display implementation
#[derive(Debug, Clone, PartialEq)]
pub enum DatabaseError {
    ConnectionFailed,
    InvalidQuery,
    AccessDenied,
    TableNotFound,
    Timeout,
}

impl fmt::Display for DatabaseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let message = match self {
            DatabaseError::ConnectionFailed => "Failed to connect to database",
            DatabaseError::InvalidQuery => "The SQL query is malformed",
            DatabaseError::AccessDenied => "Access denied: insufficient permissions",
            DatabaseError::TableNotFound => "The specified table does not exist",
            DatabaseError::Timeout => "Database operation timed out",
        };
        write!(f, "{}", message)
    }
}

impl Error for DatabaseError {}
```

I also like the idea of associating the enum with a result code for easier processing and debugging. Here's an earlier example from Claude:

```rust
// Enum-based error for multiple error types
#[derive(Debug)]
pub enum AppError {
    InvalidInput(String),
    NetworkError(String),
    DatabaseError { message: String, code: i32 },
}

impl fmt::Display for AppError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            AppError::InvalidInput(msg) => write!(f, "Invalid input: {}", msg),
            AppError::NetworkError(msg) => write!(f, "Network error: {}", msg),
            AppError::DatabaseError { message, code } => {
                write!(f, "Database error ({}): {}", code, message)
            }
        }
    }
}

impl Error for AppError {}
```

I am wondering about whether it's worth recording additional fields as well, for example in case of the Expired case. When I think about the Verification Service, I image multiple consumers: EVI, VSI and third party services. For EVI we don't need detailed error messages, for VSI and third party services it could potentially be interesting to have this additional information to help (users) understand why a verification failed.

simne7 replied:
What would be the goal of associating additional data with the enum variants? Would you like to extend the result item in a way like 

```json
{
  "documents": [{}],
  "credentials": [{},{}],
  "results": [{
    "message": "VC verified",
    "code": 100
  }, {
    "message": "VC expired",
    "code": 220,
    "expiredAt": "2024-06-01" 
  }]
}
```

simne7 replied:
Instead of introducing another discriminator for the result types like `code`  I decided to tell serde to tag each variant with the enum variant name. Manually adding a `code` field has the downside that we need to maintain the codes and have to make sure they are used uniquely. 

The result now looks like

```json
{
  "documents": [{}],
  "credentials": [{},{}],
  "results": [
    {
      "result": "VcValid",
      "message": "Verifiable Credential is valid.",
      "details": ""
    },
    {
      "result": "VcValidationErrorExpired",
      "message": "Verifiable Credential has expired.",
      "details": "2000-12-31T23:59:00+00:00"
    }
  ]
}
```

For now I used the same payload for every result item and did not introduce additional fields for single variants (like "expiredAt" for VcValidationErrorExpired). This simplifies the types and is still usable on the consumer side.

jceb reviewed (changes requested):
The PR looks good, verify nice work :clap:  I made a comment about the results format. Maybe we can make it more human + machine readable.

jceb commented on services/verification-service/src/verifications/dto.rs:
```diff
@@ -121,6 +121,25 @@ fn is_valid_url(url: &str) -> bool {
 }
 
+#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
+pub enum VcVerificationResult {
```
Should this be renamed to VerificationResult since both VP and VC verification data is represented?

simne7 replied:
done

jceb commented on services/verification-service/src/verifications/dto.rs:
```diff
@@ -60,11 +60,11 @@ impl IntoResponse for VerificationError {
     pub documents: Vec<Document>,
     pub credentials: Vec<SpecializedJsonCredential>,
-    pub results: Vec<String>,
+    pub results: Vec<VcVerificationResult>,
```
Could we also add another field just called something like `verified` of type boolean that is true if all VerificationResults are valid?

simne7 replied:
done

jceb commented on services/verification-service/src/verifications/verification_service.rs:
```diff
@@ -0,0 +1,295 @@
+/// Credentials.
+async fn verify_vp(vp_json: &String) -> Result<Vec<VcVerificationResult>, VcVerificationResult> {
+    // Create DataIntegrity from JSON string
+    let vp: AnyDataIntegrity<JsonPresentation> = match serde_json::from_str(&vp_json) {
```
Does this cover Data integrity _and_ also JWT proofs? For VDS, I also need support for JWTs as well. If this could be supported by the verification code, this would be terrific! If the rest of the VS code doesn't support JWTs yet, that's fine for this PR.

simne7 replied:
No, code expects input to be JSON. There should be a "TODO: support JWT" somewhere in the code :)

jceb commented on services/verification-service/src/verifications/service.rs:
```diff
@@ -88,6 +88,11 @@ pub async fn verify_by_url(url: &Url) -> Result<VerificationResponseDto, Error>
+
+        // verify VPs and nested VCs
+        let verification_results =
+            verification_service::verify_presentations(linked_presentations).await;
```
Is the Well-Known DID Configuration also verified?
I just set up an evil-demo-shop that has a valid DID but it just copied the did-configuration.json from the demo-shop: https://evil-demo-shop.check.identinet.io/
Currently, VS get's confused by the domain linkage credential and thinks evil-demo-shop points to the regular demo-shop's DID https://check.identinet.io/?q=https%3A%2F%2Fevil-demo-shop.check.identinet.io

simne7 replied:
I'm currently working on the DID config verification and validation (TG-190). It will 
a) verify that DID config VC is properly signed etc
b) validate that DID used in config, VP (holder) and VC (subject.id) are the same

jceb commented:
I just deployed another shop: evil2. It provides a proper DID and a signed did-configuration.json + a presentation. However, the presentation contains copied credentials from the original demo shop. So the credentialSubject.id of the credentials doesn't match the holder ID in the presentation + proof ID.

I also extended the documentation to include references to all environments: https://github.com/identinet/check?tab=readme-ov-file#demo-deployment

simne7 committed:
feat(verification-service): add verification details to results

simne7 committed:
feat(verification-service): add aggregated "verified" prop to result

simne7 committed:
feat(verification-service): verify did configuration

simne7 committed:
chore(verification-service): add generator script for did configurations

simne7 committed:
feat(verification-service): ensure presentation holder matches did

simne7 committed:
fix(verification-service): move holder-check after presentation integrity check

simne7 committed:
feat(verification-service): ensure VC subject match VP holder

jceb commented:
It's looking good :+1:  I could verify the functionality.

The one thing I'm not so happy about is the result when https://evil1-demo-shop.check.identinet.io and https://evil3-demo-shop.check.identinet.io are queried:
- the return code is 400 / Bad Request. I'd say the request wasn't bad, the result is bad. So maybe 200 / OK is better.
- the result object just contains an `error` property but no `verified` property. I'd say `error` + `verified` should be present to unify the processing of results.
- the UI should display the error message instead of the HTTP code

![image](https://github.com/user-attachments/assets/3732e799-39c1-4c27-a762-6f188cb345e0)
- the UI doesn't indicate that there's something wrong with https://evil2-demo-shop.check.identinet.io despite the API returning `verified` false. I'd say let's display each credential that failed to be validated in a different color and let's include the error message + details
![image](https://github.com/user-attachments/assets/cfd9b31b-e19d-49db-8137-38c7b161cc61)

jceb commented:
@simne7 would you take a look again at the results of this call `curl https://api.check.identinet.io/v1/verification?q=https://demo-shop.check.identinet.io`. I did deploy the latest version of VS. In the results section, there are two "Subject must be a DID" errors in the first credentials. However, when I look at the list of credentials, the first has a proper DID. The second and the third don't have a proper DID. It seems like the ordering is off.

Furthermore, when I look at the results objects, I find it difficult to see which ones have been successfully verified. Maybe we could include a boolean that indicates the verification status in each result object. BTW, the `result` property is currently a string. Did we not say that we wanted to have an integer? I don't recall what we agreed upon.

jceb commented:
Here's the result for the curl call:

```json
{
  "documents": [
    {
      "id": "did:web:demo-shop.check.identinet.io",
      "authentication": [
        "did:web:demo-shop.check.identinet.io#sDETTmw1vAJwnKQIz_CQbLzGaKoHrkf7AsEVjm6-zSg"
      ],
      "assertionMethod": [
        "did:web:demo-shop.check.identinet.io#sDETTmw1vAJwnKQIz_CQbLzGaKoHrkf7AsEVjm6-zSg"
      ],
      "verificationMethod": [
        {
          "id": "did:web:demo-shop.check.identinet.io#sDETTmw1vAJwnKQIz_CQbLzGaKoHrkf7AsEVjm6-zSg",
          "type": "JsonWebKey2020",
          "controller": "did:web:demo-shop.check.identinet.io",
          "publicKeyJwk": {
            "crv": "P-256",
            "kty": "EC",
            "x": "jUuwALI_0URe3gJmIIkOtCK493wuQbc_sQxCXvT2zBI",
            "y": "r7YU4ZHxzNXDFpto-Mk1861vLqYQ-5l_CJFI6-c2NzU"
          }
        },
        {
          "id": "did:web:demo-shop.check.identinet.io#5NTaZ3h7ayttwcd5VQ3JSqDPGlEz1IsryshZsl0JXJA",
          "type": "JsonWebKey2020",
          "controller": "did:web:demo-shop.check.identinet.io",
          "publicKeyJwk": {
            "crv": "P-256",
            "kty": "EC",
            "x": "rWOxKjGjG6j5LzIDCSKB1M8Xw99iikZZT4euQsuWrO4",
            "y": "WvF1DofkMNEBPJ-jqLwnibZN7DtVyFK5_LmgxKs_Sc0"
          }
        }
      ],
      "service": [
        {
          "id": "did:web:demo-shop.check.identinet.io#presentation",
          "type": "LinkedVerifiablePresentation",
          "serviceEndpoint": "https://demo-shop.check.identinet.io/.well-known/presentation.json"
        },
        {
          "id": "did:web:demo-shop.check.identinet.io#domain",
          "type": "LinkedDomains",
          "serviceEndpoint": [
            "https://demo-shop.check.identinet.io"
          ]
        }
      ],
      "@context": [
        "https://www.w3.org/ns/did/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
        "https://identity.foundation/.well-known/did-configuration/v1",
        "https://identity.foundation/linked-vp/contexts/v1"
      ]
    }
  ],
  "credentials": [
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
        {
          "schema": "https://schema.org/"
        }
      ],
      "type": [
        "VerifiableCredential",
        "schema:Organization"
      ],
      "credentialSubject": {
        "id": "did:web:demo-shop.check.identinet.io",
        "schema:foundingDate": "2024-11-11T11:11:11Z",
        "schema:foundingLocation": {
          "schema:address": {
            "schema:addressCountry": "DE",
            "schema:addressLocality": "Shopping City",
            "schema:addressRegion": "Shopping State",
            "schema:postalCode": "12345",
            "schema:streetAddress": "Some Street 1",
            "type": "schema:PostalAddress"
          },
          "type": "schema:Place"
        },
        "schema:legalName": "Demo Shop"
      },
      "issuer": "did:web:identinet.io",
      "issuanceDate": "2025-05-16T16:32:35Z",
      "proof": {
        "type": "JsonWebSignature2020",
        "proofPurpose": "assertionMethod",
        "verificationMethod": "did:web:identinet.io#gl7PMtYLh9ImpAetADqRjzAq3j3lqPdW_-AeRAnVMOo",
        "created": "2025-05-16T16:32:35.728977132Z",
        "jws": "eyJhbGciOiJFUzI1NiIsImtpZCI6ImdsN1BNdFlMaDlJbXBBZXRBRHFSanpBcTNqM2xxUGRXXy1BZVJBblZNT28iLCJjcml0IjpbImI2NCJdLCJiNjQiOmZhbHNlfQ..OnOCLy3L9ZoZIYPQw5AfqI7HBLNw75sbOfDns6MVOJ9hAAEmGfDKbX3Z8CaQGH9fQXsK_jeWPC1OfDqEivCBQg"
      }
    },
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
        {
          "schema": "https://schema.org/"
        }
      ],
      "type": [
        "VerifiableCredential",
        "schema:ComputerStore",
        "schema:OnlineStore"
      ],
      "credentialSubject": {
        "id": "https://demo-shop.check.identinet.io",
        "schema:vatID": "DE123456789",
        "schema:taxID": "12 345 67890",
        "schema:location": {
          "schema:addressCountry": "DE",
          "schema:addressLocality": "Shopping City",
          "schema:addressRegion": "Shopping State",
          "schema:postalCode": "12345",
          "schema:streetAddress": "Some Street 1",
          "type": "schema:PostalAddress"
        },
        "schema:openingHours": "Mo,Sa 10:00-20:00",
        "schema:telephone": "+12 3456 7890123",
        "schema:email": "support@demo-shop.emxample.com",
        "schema:priceRange": "$$",
        "schema:legalName": "Demo Shop",
        "schema:paymentAccepted": "Cash, Credit Card, Cryptocurrency"
      },
      "issuer": "did:web:demo-shop.check.identinet.io",
      "issuanceDate": "2025-05-16T16:33:10Z",
      "proof": {
        "type": "JsonWebSignature2020",
        "proofPurpose": "assertionMethod",
        "verificationMethod": "did:web:demo-shop.check.identinet.io#sDETTmw1vAJwnKQIz_CQbLzGaKoHrkf7AsEVjm6-zSg",
        "created": "2025-05-16T16:33:11.040418298Z",
        "jws": "eyJhbGciOiJFUzI1NiIsImtpZCI6InNERVRUbXcxdkFKd25LUUl6X0NRYkx6R2FLb0hya2Y3QXNFVmptNi16U2ciLCJjcml0IjpbImI2NCJdLCJiNjQiOmZhbHNlfQ..4PxiyXnIvtxQ5xOEi3I9cAd6WYLxZjkIcZAA7RPzoQeOlN4j4HqzH1FKGopR46eT9peb2aF1Srkyib5vM_1fxw"
      }
    },
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
        {
          "schema": "https://schema.org/"
        }
      ],
      "type": [
        "VerifiableCredential",
        "schema:MerchantReturnPolicy"
      ],
      "credentialSubject": {
        "id": "https://demo-shop.check.identinet.io",
        "schema:applicableCountry": "DE",
        "schema:inStoreReturnsOffered": true,
        "schema:refundType": [
          "FullRefund"
        ],
        "schema:returnFees": [
          "FreeReturn"
        ],
        "schema:returnMethod": [
          "ReturnByMail",
          "ReturnInStore"
        ],
        "schema:merchantReturnDays": 14
      },
      "issuer": "did:web:demo-shop.check.identinet.io",
      "issuanceDate": "2025-05-16T16:33:32Z",
      "proof": {
        "type": "JsonWebSignature2020",
        "proofPurpose": "assertionMethod",
        "verificationMethod": "did:web:demo-shop.check.identinet.io#sDETTmw1vAJwnKQIz_CQbLzGaKoHrkf7AsEVjm6-zSg",
        "created": "2025-05-16T16:33:33.041526501Z",
        "jws": "eyJhbGciOiJFUzI1NiIsImtpZCI6InNERVRUbXcxdkFKd25LUUl6X0NRYkx6R2FLb0hya2Y3QXNFVmptNi16U2ciLCJjcml0IjpbImI2NCJdLCJiNjQiOmZhbHNlfQ..slVW_fq7EEixinnUqhtt_isDFl8pDAdCItO0X_4Uk_VKKEBjflB9FaRUVofrW_VPnahjmoKe4AVN4hcW3bDdBA"
      }
    },
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/jws-2020/v1",
        {
          "schema": "https://schema.org/"
        }
      ],
      "type": [
        "VerifiableCredential",
        "schema:Service"
      ],
      "credentialSubject": {
        "id": "did:web:demo-shop.check.identinet.io",
        "schema:identifier": "did:web:demo-shop.check.identinet.io",
        "schema:name": "Customer Service",
        "schema:award": "Outstanding Service 2025"
      },
      "issuer": "did:web:identinet.io",
      "issuanceDate": "2025-05-21T11:37:16Z",
      "proof": {
        "type": "JsonWebSignature2020",
        "proofPurpose": "assertionMethod",
        "verificationMethod": "did:web:identinet.io#gl7PMtYLh9ImpAetADqRjzAq3j3lqPdW_-AeRAnVMOo",
        "created": "2025-05-21T11:37:16.524987048Z",
        "jws": "eyJhbGciOiJFUzI1NiIsImtpZCI6ImdsN1BNdFlMaDlJbXBBZXRBRHFSanpBcTNqM2xxUGRXXy1BZVJBblZNT28iLCJjcml0IjpbImI2NCJdLCJiNjQiOmZhbHNlfQ..Z0DdmRPHQWPSsG3SUZZ5LDawOlMnq991Lx1g28CxatAaFPv_SJe9llgInCZMiE025f07bE8b1Va8BZIcJrOytA"
      }
    }
  ],
  "results": [
    {
      "result": "VcValidationErrorOther",
      "message": "Validation failed.",
      "details": "Subject must be a DID"
    },
    {
      "result": "VcValidationErrorOther",
      "message": "Validation failed.",
      "details": "Subject must be a DID"
    },
    {
      "result": "VcValid",
      "message": "Verifiable Credential is valid.",
      "details": ""
    },
    {
      "result": "VcValid",
      "message": "Verifiable Credential is valid.",
      "details": ""
    }
  ],
  "verified": false
}
```

jceb commented:
I just fixed the issues in the VCs so the verification now succeeds. However, the VC-result ordering issue remains.

jceb commented:
Here, the error still exists: `curl https://api.check.identinet.io/v1/verification?q=https://evil2-demo-shop.check.identinet.io`

simne7 committed:
feat(verification-service): add verified prop to every result item

simne7 committed:
feat(verification-service): return 200 instead of 400 if DID config is invalid

simne7 committed:
fix(verification-service): sort verification results correctly

Before results of the verification tasks were sorted in the order they completed
not the order they were spawned. Now we sort them after the completed by the
order they were spawned.

simne7 commented:
@jceb I addressed:


[880c7d7](https://github.com/identinet/check/pull/1/commits/880c7d72ade1064e01a1da08f885ff0c9ffcd138)

No 400 Bad Request if DID config verification fails, now returns something like

```json
{
  "error": "credentialSubject.origin must match the origin the resource was requested from",
  "verified": false
}
```

`curl http://localhost:3012/v1/verification?q=https://evil-demo-shop.check.identinet.io`

[8ddb37b](https://github.com/identinet/check/pull/1/commits/8ddb37b9a1321aa8ef52e019b1ddd0a943231598)

added a verified property to each result item

```json
{
 "results": [
    {
      "result": "VcValid",
      "message": "Verifiable Credential is valid.",
      "details": "",
      "verified": true, // <<---- this one
    },
  ],
  "verified": true
}
```

`curl http://localhost:3012/v1/verification?q=https://demo-shop.check.identinet.io`


[745596d](https://github.com/identinet/check/pull/1/commits/745596d4371a6e8a7af26958e375a022d909a49f)

fixed to sort order of the results (so far they were sorted in order the verification tasks completed, not the order the tasks were spawned)

simne7 committed:
fix(verification-service-ui): ensure result card has width of column

simne7 committed:
feat(verification-service-ui): show error message from verification result

simne7 committed:
feat(verification-service-ui): render invalid vcs correctly

jceb committed:
fix(verification-service-ui): cleanup linter errors

jceb committed:
fix(embedded-verification-ui): correct type annotations

jceb committed:
fix(verification-service-ui): correct award handling

jceb committed:
fix(verification-service-ui): change credential layout to flexbox

jceb committed:
chore(verification-service-ui): update dependencies

jceb committed:
chore(verification-service-ui): cleanup unused files

jceb committed:
feat(verification-service-ui): rework UI

jceb commented on services/verification-service-ui/src/components/CredentialCard.tsx:
```diff
@@ -1,150 +1,150 @@
   message: string;
 } & JSX.HTMLAttributes<HTMLDivElement>;
 
-export const NotFoundCard: Component<Props> = ({ icon, message }) => {
```
@simne7 FYI, reactivity breaks in solid-js when props are destructured.
